### PR TITLE
Rendering code

### PR DESCRIPTION
**Pull request by dtracers**
Allows  bots to render what they would like to render

### CODE REVIEW

File: src/main/flatbuffers/rlbot.fbs [hunk 1]
```diff
@@ -185,4 +185,43 @@ table DesiredGameState {
...
+    DrawLine2D = 1,
+    DrawLine3D = 2,
+    DrawLine2D_3D = 3,
+    DrawRect2D = 4,
+    DrawRect3D = 5,
+    DrawString2D = 6,
+    DrawString3D = 7
+}
+
+table Color {
```
Comment (tarehart) [inline src/main/flatbuffers/rlbot.fbs]: This is unlikely to change, so you could probably use a struct. 

> structs are ideal for data structures that will not change, since they use less memory and have faster lookup

File: src/main/python/RLBotFramework/utils/rendering/rendering_manager.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    ignored_funcs = []
+
+    def __init__(self):
+        self.ignored_funcs = ['setup_function_types', 'get_render_functions',
+                              'create_dynamic_function', 'send_group']
+
+    def setup_function_types(self, dll_instance):
+        self.game = dll_instance
+
+        func = self.game.DrawLine2D
```
Comment (tarehart) [inline src/main/python/RLBotFramework/utils/rendering/rendering_manager.py]: Why is this needed? Don't we use RenderGroup for everything?

File: src/main/python/RLBotFramework/utils/rendering/rendering_manager.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+    def black(self):
+        return self.create_color(255, 0, 0, 0)
+
+    def white(self):
+        return self.create_color(255, 255, 255, 255)
+
+    def get_render_functions(self):
+        """
+        Gets all the raw render functions but without giving access to any internal logic or the dll
```
Comment (tarehart) [inline src/main/python/RLBotFramework/utils/rendering/rendering_manager.py]: Why not just use the `__` prefix for private stuff? That should be enough to prevent accidental misuse. Is there some benefit to this dynamic function technique that I'm missing?

File: src/test/python/agents/atba/atba.py [hunk 4]
```diff
@@ -33,6 +33,24 @@ def get_output_vector(self, game_tick_packet):
         if turn == -1.0:
             self.send_quick_chat(QuickChats.CHAT_EVERYONE, QuickChats.Information_IGotIt)
 
+        car_render_location = [my_car.Location.X, my_car.Location.Y, my_car.Location.Z]
+        ball_render_location = [game_tick_packet.gameball.Location.X,
+                                game_tick_packet.gameball.Location.Y,
+                                game_tick_packet.gameball.Location.Z]
+        text_render_strX = 'x:' + str(game_tick_packet.gameball.Location.X)
+        self.renderer.begin_rendering()
```
Comment (tarehart) [inline src/test/python/agents/atba/atba.py]: Why expose begin_rendering and end_rendering to the bot? Can't the bot manager handle that, and just no-op during end_rendering if the bot never called any draw functions?

Comment (dtracers) [reply src/test/python/agents/atba/atba.py]: I would rather have the bot makers have control over what group they want to render in case they want multiple groups that can asynchronously render.

File: src/main/python/RLBotFramework/utils/rendering/rendering_manager.py (hunk 3 above)
Comment (tarehart) [reply src/main/python/RLBotFramework/utils/rendering/rendering_manager.py]: A bot could also construct an additional instance of the game interface without any help from the rendering system. Malicious tournament players have a ton of options, and we're pretty much on the honor system given the current architecture.

File: rlbot.cfg [hunk 5]
```diff
@@ -12,8 +12,8 @@ game_map = 34
 [Participant Configuration]
 # Put the name of your bot config file here.  Only total_num_participants config files will be read!
 # Everything needs a config, even players and default bots.  We still set loadouts and names from config!
-participant_config_0 = src/test/python/agents/flatBot/flatBot.cfg
-participant_config_1 = src/test/python/agents/atba/atba.cfg
+participant_config_1 = src/test/python/agents/flatBot/flatBot.cfg
```
[the remark this answers is not in the document]
Comment (dtracers) [reply rlbot.cfg]: done

File: src/main/cpp/RLBotInterface/RLBotInterface/BoostUtilities/BoostUtilities.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#include <boost\interprocess\ipc\message_queue.hpp>
+#include <boost/interprocess/shared_memory_object.hpp>
+#include <boost/interprocess/mapped_region.hpp>
+#include <boost\interprocess\sync\named_sharable_mutex.hpp>
+#include <boost\interprocess\sync\sharable_lock.hpp>
+
+#include "BoostUtilities.hpp"
+
+
+namespace GameFunctions {
```
[the remark this answers is not in the document]
Comment (dtracers) [reply src/main/cpp/RLBotInterface/RLBotInterface/BoostUtilities/BoostUtilities.cpp]: done

File: src/main/cpp/RLBotInterface/RLBotInterface/BoostUtilities/BoostUtilities.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
+#ifndef BOOSTUTILITIES_HPP
```
[the remark this answers is not in the document]
Comment (dtracers) [reply src/main/cpp/RLBotInterface/RLBotInterface/BoostUtilities/BoostUtilities.hpp]: done

File: src/main/cpp/RLBotInterface/RLBotInterface/RenderFunctions/RenderFunctions.cpp [hunk 8]
```diff
@@ -118,4 +118,16 @@ namespace RenderFunctions
 
 		return RLBotCoreStatus::Success;
 	}
-}
\ No newline at end of file
+
+	// Currently we are relying on the core dll to create the queue in shared memory before this process starts. TODO: be less fragile
+	static interop_message_queue renderGroupQueue(boost::interprocess::open_only, BoostConstants::RenderingFlatQueueName);
```
[the remark this answers is not in the document]
Comment (dtracers) [reply src/main/cpp/RLBotInterface/RLBotInterface/RenderFunctions/RenderFunctions.cpp]: Done

File: src/main/cpp/RLBotInterface/RLBotMessages/MessageProcessor/MessageProcessor.hpp [hunk 9]
```diff
@@ -52,20 +62,15 @@ class MessageProcessor
 			it++
 			)
 		{
-			auto result = messageHandlerMap.find(it->Type);
+			//ToDo: Execute only if the callback buffer is not overfilled?
+			RLBotCoreStatus returnValue = callMessage(it->Type, it.GetCurrentMessage());
 
-			if (result != messageHandlerMap.end())
+			if (returnValue != RLBotCoreStatus::InvalidGameValues && it->HasCallback)
```
[the remark this answers is not in the document]
Comment (dtracers) [reply src/main/cpp/RLBotInterface/RLBotMessages/MessageProcessor/MessageProcessor.hpp]: Done

File: src/main/flatbuffers/rlbot.fbs (hunk 1 above)
Comment (dtracers) [reply src/main/flatbuffers/rlbot.fbs]: I left this as a table so that it can be created once and used multiple times,
It appears that is not possible with structs

File: src/main/python/RLBotFramework/utils/rendering/rendering_manager.py (hunk 2 above)
Comment (dtracers) [reply src/main/python/RLBotFramework/utils/rendering/rendering_manager.py]: Removed

File: src/main/python/RLBotFramework/utils/rendering/rendering_manager.py (hunk 3 above)
Comment (dtracers) [reply src/main/python/RLBotFramework/utils/rendering/rendering_manager.py]: Done
